File: 1428-jump-game-iii/jump-game-iii.py
from collections import deque

class Solution(object):
    def canReach(self, arr, start):
        n = len(arr)
        visited = set()

        q = deque([start])

        while q:
            i = q.popleft()

            if arr[i] == 0:
                return True

            if i in visited:
                continue

            visited.add(i)

            for nxt in (i + arr[i], i - arr[i]):
                if 0 <= nxt < n and nxt not in visited:
                    q.append(nxt)

        return False

File: 1428-jump-game-iii/jump-game-iii.dart
class Solution {
  bool canReach(List<int> arr, int start) {
    List<bool> visited = List.filled(arr.length, false);

    List<int> queue = [start];
    visited[start] = true;

    while (queue.isNotEmpty) {
      int i = queue.removeAt(0);

      if (arr[i] == 0) {
        return true;
      }

      int forward = i + arr[i];
      int backward = i - arr[i];

      if (forward >= 0 &&
          forward < arr.length &&
          !visited[forward]) {
        visited[forward] = true;
        queue.add(forward);
      }

      if (backward >= 0 &&
          backward < arr.length &&
          !visited[backward]) {
        visited[backward] = true;
        queue.add(backward);
      }
    }

    return false;
  }
}